### DISCUSSION TITLE
How to adjust eink contrast so that it looks nice on eink screen

### DISCUSSION THREAD

**Discussion by mergen3107** [koreader/koreader#11133] [General]
Hey guys,

I am sorry, this is a little off-topic question with regards KOReader. 

https://www.reddit.com/r/Onyx_Boox/comments/w9zili/nasa_posters_adapted_for_eink/

Here is an example right away. Some time ago I converted color posters for eink by adding dithering and making them with 8-bit color depth.

Then I experimented and found out that about +20% "brightness" correction looked much better on eink screen of my Onyx Boox. 

However, I don't know the correct way to do this. I know for DSLR cameras, for example, the balance between shutter and exposure must be achieved so the photos aren't too bright or too dark. 

How should I calibrate converted 16-shade of gray images to eink in the same way?

I mean when is it "too bright"?

**Comment by NiLuJe** [comment]
There's no real silver bullet if the source is difficult. Otherwise, it's as "simple" as setting a correct white point and a correct black point (i.e., sane levels).

**Reply by mergen3107** [reply]
Thank you @NiLuJe !
I am going to try with these maps…

https://www.nexusmods.com/kingdomcomedeliverance/mods/827/?tab=files

Gonna be a challenge, I suppose :D 

Here is the tutorial I found:

https://photoscreenprint.com/setting-black-white-points/

**Reply by mergen3107** [reply]
@NiLuJe 
Does Kindle Scribe support hardware dithering in KOReader?

**Reply by mergen3107** [reply]
Also I am a little lost with what order should I follow. Manga originals I get are quite smaller than 2480x1860 px on KS. I get around 800 x 1150, sometimes if lucky 1100 x 1600. So I am going to use waifu2x upscaler + dithering + KCC (but only thing I need really is to rotate double-spreads, so I'll probably add it somewhere to dithering batch file).

I have two choices:
1) Upscale those small originals 2x, then stretch-rotate with KCC to full 2480x1860, then do dithering.
2) Take my old 1872x1404 prepared mangas and just upscaled them by 1.325 to get 2480x1869. However, in this case the dithering quality will be lower, right? Because it will stretch the old dithering.

Any help would be appreciated :D

**Reply by mergen3107** [reply]
> Also I am a little lost with what order should I follow. 

I figured this out as well. 

1) Upscale manga with waifu2x: 2X scale, low noise reduction, output as PNG. 
2) Use latest Kindle Comic Converter: MAX screen resolution for target device, gamma forced to 1.00, keep color, stretch/upscale, force PNG, so that it adds white/black borders without distorting the original manga, but does nothing else to the pictures. Output as CBZ.
3) Unpack resulting CBZ, apply dithering.
Pack back. 

waifu2x is a seriously concerning great tool that uses Nvidia CUDA cores to losslessly upscale stuff. 

I mean, the pictures looks as if they were sharpened, but no artifacts are added. Another "fog" situation. 

With original resolution of ~1100x1600 px the resulting 1860x2480 px looks great! Anything lower on the input side is meh (especially those 600x1100 px ones…).

**Reply by mergen3107** [reply]
But the question remains: should I keep doing dithering by script, or Kindle Scribe is able to dither via HW (SW?) in KOReader? 

Can't test it quite yet myself, hope to be able to soon…

**Comment by mergen3107** [comment]
> How should I calibrate converted 16-shade of gray images to eink in the same way?
> 
> I mean when is it "too bright"?

I think I figured it out.
First I used the script to convert to 16-gray scale with dithering.
Then using GIMP I adjusted color levels:
1) Black point and white points as the bonding points where most of the "stuff" on the histogram starts and ends, respectively.
2) Clamp point in the middle: adjusted closer to the "bright" side, but not too much.
While clamping, I kinda moved back and forth between "too white" and "too black" - to stay somewhere in the middle, e.g., "neutral tone". I think that is what makes it look great on the eink, because eink has a hard time trying to present anything too bright or too dark (for example, try distinguishing between purple vs black color).

See picture below:
![color_levels_explained](https://github.com/koreader/koreader/assets/6279855/70e45543-18ae-456e-a090-18893a75187d)

Attached are examples of before and after color levels modification. Looks absolutely clear on the eink (tested on Onyx so far)!

[color_levels_explained.zip](https://github.com/koreader/koreader/files/13422452/color_levels_explained.zip)

The difference between "before" and "after" looks like a "fog" was removed, lol!
In Gimp, there is a handy tool "Split view" that lets you also drag the before/after boundary to see the result.

However, yes, it is quite an art at this point - this mod is done by eye, and heavily depends on the source image color distribution.

**Comment by NiLuJe** [comment]
> @NiLuJe Does Kindle Scribe support hardware dithering in KOReader?

Unknown, it didn't work on the PW5, but it might there?

**Reply by NiLuJe** [reply]
As for SW dithering, that's a toggle in the bottom menu in a PDF/CBZ, so, dealer's choice ;).

**Reply by mergen3107** [reply]
O_o I wanna try this asappppp :D 
Thanks @NiLuJe !

**Comment by NiLuJe** [comment]
> First I used the script to convert to 16-gray scale with dithering.
> Then using GIMP I adjusted color levels:

It's going to be less destructive on manga to do it *after*, but, as a rule, fixing levels is one of the *first* things I'd do, as it'll affect dithering (and possibly grayscaling depending on how much you need to fiddle).

**Reply by mergen3107** [reply]
Oh, OK! I'll keep this in mind :) 
For now, I color-correct only screensavers. I don't think I have enough sanity to do this by hand for all manga, lol. Maybe I'll do gimp cli batch file later, but I don't know to adjust them automatically yet.

**Reply by Frenzie** [reply]
Yeah, this was going to be my comment too. If you dither first and then adjust levels you're undoing the dithering, making it less nice. ;-)

**Reply by mergen3107** [reply]
Thank you guys!
I really appreciate KOReader community because you guys rock in so many areas :D

**Reply by mergen3107** [reply]
I tried to:
1) Take original color picture, color correct that to the same logic
2) Dither it.

The output is still "dimmer" that `dither + correct`:
![color first](https://github.com/koreader/koreader/assets/6279855/dbfbdaf6-4dbd-41dc-ac03-e5f2d6d45952)

I then also tried to dither my first attempts (so full order is `dither + correct + dither`).
I cannot see visual changes at all, but the size shrinked from 4 MB to 2 MB.
The newly dithered picture, I hope, should be OK for Kindle's framebuffer :D

**Reply by NiLuJe** [reply]
Basically: color correct (and any other editing you might want to do) on the original image => scale => grayscale => dither.

Dithering is extremely destructive, so needs to be at the tail-end of the pipeline.

**Reply by mergen3107** [reply]
Thank you!

For now I am adding another color correct after grayscale, because I can't color correct color picture with the same outcome as I color correct grayscale one :D maybe I am color blind, who knows :D
